File: cli/src/args.ts
import { parseArgs as parseNodeArgs } from 'node:util';
import {
  ALL_ENGINES,
  AUTO_SUMMARIZER,
  CLAUDE_PERMISSION_MODES,
  CODEX_SANDBOX_MODES,
  DEFAULT_ITERATIONS,
  DEFAULT_MAX_MEMBER_CHARS,
  DEFAULT_TEAM_SIZE,
  DEFAULT_TIMEOUT_MS,
  EFFORT_LEVELS,
  PROVIDER_AUTH_METHODS,
  PROVIDER_EFFORT_LEVELS
} from './engines.js';
const OPTIONS = {
  help: { type: 'boolean', short: 'h' },
  version: { type: 'boolean', short: 'v' },
  json: { type: 'boolean' },
  'json-stream': { type: 'boolean' },
  ndjson: { type: 'boolean' },
  headless: { type: 'boolean' },
  studio: { type: 'boolean' },
  plain: { type: 'boolean' },
  banner: { type: 'boolean', default: true },
  'summary-only': { type: 'boolean' },
  quiet: { type: 'boolean', short: 'q' },
  verbose: { type: 'boolean', short: 'd' },
  all: { type: 'boolean' },
  codex: { type: 'boolean', default: true },
  claude: { type: 'boolean', default: true },
  gemini: { type: 'boolean', default: true },
  members: { type: 'string' },
  summarizer: { type: 'string' },
  effort: { type: 'string' },
  'codex-model': { type: 'string' },
  'claude-model': { type: 'string' },
  'gemini-model': { type: 'string' },
  'codex-effort': { type: 'string' },
  'claude-effort': { type: 'string' },
  'gemini-effort': { type: 'string' },
  'codex-sandbox': { type: 'string' },
  'claude-permission-mode': { type: 'string' },
  'codex-auth': { type: 'string' },
  'claude-auth': { type: 'string' },
  'gemini-auth': { type: 'string' },
  handoff: { type: 'boolean' },
  lead: { type: 'string' },
  planner: { type: 'string' },
  iterations: { type: 'string' },
  'team-work': { type: 'string' },
  teamwork: { type: 'string' },
  'sub-agents': { type: 'string' },
  'codex-sub-agents': { type: 'string' },
  'claude-sub-agents': { type: 'string' },
  'gemini-sub-agents': { type: 'string' },
  linear: { type: 'boolean' },
  'deliver-linear': { type: 'boolean' },
  'linear-issue': { type: 'string' },
  'linear-query': { type: 'string' },
  'linear-team': { type: 'string' },
  'linear-state': { type: 'string' },
  'linear-assignee': { type: 'string' },
  'linear-limit': { type: 'string' },
  'linear-endpoint': { type: 'string' },
  'linear-api-key-env': { type: 'string' },
  'delivery-phases': { type: 'string' },
  timeout: { type: 'string' },
  'max-member-chars': { type: 'string' },
  cwd: { type: 'string' },
  color: { type: 'string' },
  'no-color': { type: 'boolean' }
} as any;

export function usageText(version) {
  const defaultTimeoutSeconds = DEFAULT_TIMEOUT_MS / 1_000;

  return [
    `council v${version}`,
    '',
    'Ask multiple AI CLIs the same question, then synthesize their answers.',
    '',
    'Usage:',
    '  council [options] <query>',
    '  echo "query" | council [options]',
    '  council help',
    '',
    'Examples:',
    '  council "How should I structure this migration?"',
    '  council --no-gemini --summarizer claude "Review this plan"',
    '  council --headless --json "Summarize the implementation options"',
    '  council --json-stream --codex --claude "Compare these designs"',
    '  council --studio "Open a configurable terminal workbench"',
    '  council --effort high "Analyze the tradeoffs for this architecture"',
    '  council --planner codex --lead claude --handoff --iterations 2 "Plan and execute this change"',
    '  council --deliver-linear --linear-issue ABC-123 --lead codex',
    '',
    'Selection:',
    '  --members <list>          Ordered subset of codex,claude,gemini',
    '  --codex / --no-codex      Enable or disable Codex',
    '  --claude / --no-claude    Enable or disable Claude',
    '  --gemini / --no-gemini    Enable or disable Gemini',
    '  --all                     Re-enable all members',
    '  --summarizer <name>       auto, codex, claude, or gemini',
    '',
    'Output:',
    '  --summary-only            Print only the final synthesis',
    '  -q, --quiet               Alias for --summary-only',
    '  -d, --verbose             Show all member responses, even if they fail',
    '  --json                    Print structured JSON',
    '  --json-stream             Stream JSONL lifecycle events',
    '  --headless                Automation mode: no banner, no progress, summary-only text by default',
    '  --studio                  Open the interactive workbench with focusable panes and editable settings',
    '  --plain                   Disable decorative UI and color',
    '  --no-banner               Suppress the startup banner',
    '  --color <auto|always|never>',
    '',
    'Execution:',
    `  --timeout <seconds>       Per-CLI timeout in seconds (default: ${defaultTimeoutSeconds} / 10 minutes)`,
    `  --max-member-chars <n>    Cap each member response before summarization (default: ${DEFAULT_MAX_MEMBER_CHARS})`,
    '  --cwd <path>              Working directory for all upstream CLIs',
    `  --effort <level>          Reasoning effort applied to every member: ${EFFORT_LEVELS.join(', ')}`,
    '  --codex-model <model>     Model passed to Codex via --model',
    '  --claude-model <model>    Model passed to Claude via --model',
    '  --gemini-model <model>    Model passed to Gemini via --model',
    `  --codex-effort <level>    Codex effort: ${PROVIDER_EFFORT_LEVELS.codex.join(', ')}`,
    `  --claude-effort <level>   Claude effort: ${PROVIDER_EFFORT_LEVELS.claude.join(', ')}`,
    `  --gemini-effort <level>   Gemini effort: ${PROVIDER_EFFORT_LEVELS.gemini.join(', ')}`,
    `  --codex-sandbox <mode>    Codex sandbox: ${CODEX_SANDBOX_MODES.join(', ')}`,
    `  --claude-permission-mode <mode>`,
    `                            Claude permission mode: ${CLAUDE_PERMISSION_MODES.join(', ')}`,
    `  --codex-auth <method>     Codex auth preference: ${PROVIDER_AUTH_METHODS.codex.join(', ')}`,
    `  --claude-auth <method>    Claude auth preference: ${PROVIDER_AUTH_METHODS.claude.join(', ')}`,
    `  --gemini-auth <method>    Gemini auth preference: ${PROVIDER_AUTH_METHODS.gemini.join(', ')}`,
    '  --handoff                 Run members in order and pass each response to the next member',
    '  --lead <name>             Lead model for final synthesis preference: codex, claude, or gemini',
    '  --planner <name>          Planner model that runs before executor members',
    `  --iterations <n>          Consultation rounds per prompt (default: ${DEFAULT_ITERATIONS})`,
    `  --team-work <n>           Let each provider coordinate up to n internal sub-agents (default: ${DEFAULT_TEAM_SIZE})`,
    '  --codex-sub-agents <n>    Codex-specific team size override',
    '  --claude-sub-agents <n>   Claude-specific team size override',
    '  --gemini-sub-agents <n>   Gemini-specific team size override',
    '  --deliver-linear          Fetch Linear tasks and run delivery phases for each task',
    '  --linear-issue <ids>      Comma-separated Linear issue IDs/keys to deliver',
    '  --linear-query <text>     Fetch matching Linear issues when explicit IDs are not provided',
    '  --linear-team <key>       Restrict Linear fetches to a team key',
    '  --linear-state <name>     Restrict Linear fetches to a state name',
    '  --linear-assignee <text>  Restrict Linear fetches to an assignee name/email',
    '  --linear-limit <n>        Max Linear issues to fetch (default: 3)',
    '  --delivery-phases <list>  Comma-separated plan,implement,verify,ship',
    '',
    'Other:',
    '  -h, --help                Show help',
    '  -v, --version             Show version',
    '',
    'Environment:',
    '  COUNCIL_CODEX_BIN         Override the codex executable path',
    '  COUNCIL_CLAUDE_BIN        Override the claude executable path',
    '  COUNCIL_GEMINI_BIN        Override the gemini executable path',
    '  LINEAR_API_KEY            Linear API key used by --deliver-linear',
    '  CLAUDE_CODE_OAUTH_TOKEN   Use Claude OAuth-token auth (omits Claude --bare mode)',
    '  ANTHROPIC_API_KEY         Use Claude API-key auth (keeps Claude --bare mode)',
    '  CLAUDE_CODE_EFFORT_LEVEL  Claude effort fallback when no Claude effort flag is set'
  ].join('\n');
}

export function parseArgs(argv) {
  const parsed = parseNodeArgs({
    args: argv,
    options: OPTIONS,
    allowPositionals: true,
    allowNegative: true,
    strict: true,
    tokens: true
  });

  const values = parsed.values as any;
  const result = {
    help: Boolean(values.help),
    version: Boolean(values.version),
    json: Boolean(values.json),
    jsonStream: Boolean(values['json-stream'] || values.ndjson),
    headless: Boolean(values.headless),
    studio: Boolean(values.studio),
    plain: Boolean(values.plain),
    banner: Boolean(values.banner),
    summaryOnly: Boolean(values['summary-only']),
    quiet: Boolean(values.quiet),
    verbose: Boolean(values.verbose),
    noBanner: values.banner === false,
    color: values['no-color'] ? 'never' : parseColor(values.color ?? 'auto'),
    summarizer: parseSummarizer(values.summarizer ?? AUTO_SUMMARIZER),
    effort: parseEffort(values.effort),
    models: parseProviderModels(values),
    efforts: parseProviderEfforts(values),
    permissions: parseProviderPermissions(values),
    auths: parseProviderAuths(values),
    handoff: Boolean(values.handoff),
    lead: parseOptionalEngine(values.lead, '--lead'),
    planner: parseOptionalEngine(values.planner, '--planner'),
    iterations: values.iterations
      ? parsePositiveInteger(values.iterations, '--iterations')
      : DEFAULT_ITERATIONS,
    teamWork: parseTeamWork(values),
    teams: parseProviderTeams(values),
    delivery: parseDeliveryOptions(values),
    timeoutMs: values.timeout
      ? parseTimeoutMs(values.timeout)
      : DEFAULT_TIMEOUT_MS,
    maxMemberChars: values['max-member-chars']
      ? parsePositiveInteger(values['max-member-chars'], '--max-member-chars')
      : DEFAULT_MAX_MEMBER_CHARS,
    cwd: values.cwd ?? process.cwd(),
    promptParts: [...parsed.positionals]
  } as any;

  const enabledEngines = {
    codex: true,
    claude: true,
    gemini: true
  };
  let memberOrder = [...ALL_ENGINES];

  for (const token of parsed.tokens) {
    if (token.kind !== 'option') {
      continue;
    }

    if (token.name === 'all') {
      for (const engine of ALL_ENGINES) {
        enabledEngines[engine] = true;
      }
      continue;
    }

    if (token.name === 'members') {
      const members = parseEngineList(token.value, '--members');
      applyEngineList(enabledEngines, members);
      memberOrder = buildMemberOrder(members);
      continue;
    }

    if (ALL_ENGINES.includes(token.name)) {
      enabledEngines[token.name] = !token.rawName.startsWith('--no-');
    }
  }

  if (result.help || isStandaloneSubcommand(argv, 'help')) {
    result.help = true;
    result.members = [...ALL_ENGINES];
    return result;
  }

  if (result.version || isStandaloneSubcommand(argv, 'version')) {
    result.version = true;
    result.members = [...ALL_ENGINES];
    return result;
  }

  result.members = memberOrder.filter((engine) => enabledEngines[engine]);

  if (result.members.length === 0) {
    throw new Error('At least one engine must be enabled.');
  }

  validateSelectedRole(result.lead, '--lead', result.members);
  validateSelectedRole(result.planner, '--planner', result.members);

  return result;
}

function parseEngineList(value, flagName) {
  const members = value
    .split(',')
    .map((member) => member.trim())
    .filter(Boolean);

  if (members.length === 0) {
    throw new Error(`${flagName} requires at least one engine.`);
  }

  const invalid = members.filter((member) => !ALL_ENGINES.includes(member));
  if (invalid.length > 0) {
    throw new Error(`Unsupported engine in ${flagName}: ${invalid.join(', ')}`);
  }

  return [...new Set(members)];
}

function applyEngineList(enabledEngines, members) {
  for (const engine of ALL_ENGINES) {
    enabledEngines[engine] = false;
  }

  for (const member of members) {
    enabledEngines[member] = true;
  }
}

function buildMemberOrder(members) {
  return [
    ...members,
    ...ALL_ENGINES.filter((engine) => !members.includes(engine))
  ];
}

function parseSummarizer(value) {
  if (value === AUTO_SUMMARIZER || ALL_ENGINES.includes(value)) {
    return value;
  }

  throw new Error(`Unsupported summarizer: ${value}`);
}

function parseEffort(value) {
  if (value === undefined || value === null || value === '') {
    return null;
  }

  if (EFFORT_LEVELS.includes(value)) {
    return value;
  }

  throw new Error(
    `Unsupported --effort value: ${value} (expected ${EFFORT_LEVELS.join(', ')})`
  );
}

function parseProviderModels(values) {
  return {
    codex: parseOptionalString(values['codex-model'], '--codex-model'),
    claude: parseOptionalString(values['claude-model'], '--claude-model'),
    gemini: parseOptionalString(values['gemini-model'], '--gemini-model')
  };
}

function parseProviderEfforts(values) {
  return {
    codex: parseProviderEffort(values['codex-effort'], 'codex'),
    claude: parseProviderEffort(values['claude-effort'], 'claude'),
    gemini: parseProviderEffort(values['gemini-effort'], 'gemini')
  };
}

function parseProviderPermissions(values) {
  return {
    codex: parseEnumValue(values['codex-sandbox'], '--codex-sandbox', CODEX_SANDBOX_MODES),
    claude: parseEnumValue(
      values['claude-permission-mode'],
      '--claude-permission-mode',
      CLAUDE_PERMISSION_MODES
    ),
    gemini: null
  };
}

function parseProviderAuths(values) {
  return {
    codex: parseEnumValue(values['codex-auth'], '--codex-auth', PROVIDER_AUTH_METHODS.codex),
    claude: parseEnumValue(values['claude-auth'], '--claude-auth', PROVIDER_AUTH_METHODS.claude),
    gemini: parseEnumValue(values['gemini-auth'], '--gemini-auth', PROVIDER_AUTH_METHODS.gemini)
  };
}

function parseDeliveryOptions(values) {
  const enabled = Boolean(values['deliver-linear'] || values.linear);
  return {
    enabled,
    provider: enabled ? 'linear' : null,
    issueIds: parseOptionalList(values['linear-issue']),
    query: parseOptionalString(values['linear-query'], '--linear-query'),
    team: parseOptionalString(values['linear-team'], '--linear-team'),
    state: parseOptionalString(values['linear-state'], '--linear-state'),
    assignee: parseOptionalString(values['linear-assignee'], '--linear-assignee'),
    limit: values['linear-limit']
      ? parsePositiveInteger(values['linear-limit'], '--linear-limit')
      : 3,
    endpoint: parseOptionalString(values['linear-endpoint'], '--linear-endpoint'),
    apiKeyEnv: parseOptionalString(values['linear-api-key-env'], '--linear-api-key-env') || 'LINEAR_API_KEY',
    phases: parseDeliveryPhases(values['delivery-phases'])
  };
}

function parseOptionalList(value) {
  if (value === undefined || value === null || value === '') {
    return [];
  }

  return String(value)
    .split(',')
    .map((item) => item.trim())
    .filter(Boolean);
}

function parseDeliveryPhases(value) {
  const phases = parseOptionalList(value || 'plan,implement,verify,ship');
  const allowed = ['plan', 'implement', 'verify', 'ship'];
  const invalid = phases.filter((phase) => !allowed.includes(phase));

  if (invalid.length > 0) {
    throw new Error(`Unsupported --delivery-phases value: ${invalid.join(', ')} (expected ${allowed.join(', ')})`);
  }

  return phases;
}

function parseTeamWork(values) {
  const value = values['team-work'] ?? values.teamwork ?? values['sub-agents'];
  return value === undefined || value === null || value === ''
    ? DEFAULT_TEAM_SIZE
    : parseNonNegativeInteger(value, '--team-work');
}

function parseProviderTeams(values) {
  return {
    codex: parseOptionalNonNegativeInteger(values['codex-sub-agents'], '--codex-sub-agents'),
    claude: parseOptionalNonNegativeInteger(values['claude-sub-agents'], '--claude-sub-agents'),
    gemini: parseOptionalNonNegativeInteger(values['gemini-sub-agents'], '--gemini-sub-agents')
  };
}

function parseProviderEffort(value, engine) {
  if (value === undefined || value === null || value === '') {
    return null;
  }

  const allowed = PROVIDER_EFFORT_LEVELS[engine];
  if (allowed.includes(value)) {
    return value;
  }

  throw new Error(
    `Unsupported --${engine}-effort value: ${value} (expected ${allowed.join(', ')})`
  );
}

function parseEnumValue(value, flagName, allowed) {
  if (value === undefined || value === null || value === '') {
    return null;
  }

  if (allowed.includes(value)) {
    return value;
  }

  throw new Error(
    `Unsupported ${flagName} value: ${value} (expected ${allowed.join(', ')})`
  );
}

function parseOptionalString(value, flagName) {
  if (value === undefined || value === null) {
    return null;
  }

  const parsed = String(value).trim();
  if (parsed) {
    return parsed;
  }

  throw new Error(`${flagName} requires a non-empty value.`);
}

function parseOptionalEngine(value, flagName) {
  if (value === undefined || value === null || value === '') {
    return null;
  }

  const parsed = String(value).trim();
  if (ALL_ENGINES.includes(parsed)) {
    return parsed;
  }

  throw new Error(`Unsupported ${flagName} value: ${parsed} (expected ${ALL_ENGINES.join(', ')})`);
}

function validateSelectedRole(value, flagName, members) {
  if (!value || members.includes(value)) {
    return;
  }

  throw new Error(`${flagName} must be one of the enabled members: ${members.join(', ')}`);
}

function parseTimeoutMs(value) {
  const seconds = Number(value);

  if (!Number.isFinite(seconds) || seconds <= 0) {
    throw new Error(`Invalid timeout value: ${value}`);
  }

  return Math.round(seconds * 1000);
}

function parsePositiveInteger(value, flagName) {
  const parsed = Number(value);

  if (!Number.isInteger(parsed) || parsed <= 0) {
    throw new Error(`${flagName} requires a positive integer.`);
  }

  return parsed;
}

function parseOptionalNonNegativeInteger(value, flagName) {
  if (value === undefined || value === null || value === '') {
    return null;
  }

  return parseNonNegativeInteger(value, flagName);
}

function parseNonNegativeInteger(value, flagName) {
  const parsed = Number(value);

  if (!Number.isInteger(parsed) || parsed < 0) {
    throw new Error(`${flagName} requires a non-negative integer.`);
  }

  return parsed;
}

function parseColor(value) {
  if (['auto', 'always', 'never'].includes(value)) {
    return value;
  }

  throw new Error(`Unsupported color mode: ${value}`);
}

function isStandaloneSubcommand(argv, target) {
  const positionalTokens = [];

  for (const token of argv) {
    if (token === '--') {
      return false;
    }

    if (!token.startsWith('-')) {
      positionalTokens.push(token);
    }
  }

  return positionalTokens.length === 1 && positionalTokens[0] === target;
}

File: cli-rs/src/args.rs
use clap::{ArgAction, Parser};

pub const ALL_ENGINES: &[&str] = &["codex", "claude", "gemini"];
pub const AUTO_SUMMARIZER: &str = "auto";
pub const EFFORT_LEVELS: &[&str] = &["low", "medium", "high"];
pub const DEFAULT_TIMEOUT_MS: u64 = 600_000;
pub const DEFAULT_MAX_MEMBER_CHARS: usize = 12_000;

#[derive(Parser, Debug)]
#[command(
    name = "council",
    about = "Ask multiple AI CLIs the same question, then synthesize their answers.",
    long_about = None,
    disable_version_flag = true,
)]
pub struct RawArgs {
    /// Per-CLI timeout in seconds (default: 600 / 10 minutes)
    #[arg(long)]
    pub timeout: Option<String>,

    /// Cap each member response before summarization (default: 12000)
    #[arg(long = "max-member-chars")]
    pub max_member_chars: Option<String>,

    /// Working directory for all upstream CLIs
    #[arg(long)]
    pub cwd: Option<String>,

    /// Print structured JSON result at end
    #[arg(long, action = ArgAction::SetTrue)]
    pub json: bool,

    /// Stream JSONL lifecycle events as they happen
    #[arg(long = "json-stream", action = ArgAction::SetTrue)]
    pub json_stream: bool,

    /// Alias for --json-stream
    #[arg(long, action = ArgAction::SetTrue)]
    pub ndjson: bool,

    /// Automation mode: no banner, no progress, summary-only text by default
    #[arg(long, action = ArgAction::SetTrue)]
    pub headless: bool,

    /// Disable decorative UI and color
    #[arg(long, action = ArgAction::SetTrue)]
    pub plain: bool,

    /// Suppress the startup banner
    #[arg(long = "no-banner", action = ArgAction::SetTrue)]
    pub no_banner: bool,

    /// Print only the final synthesis
    #[arg(long = "summary-only", action = ArgAction::SetTrue)]
    pub summary_only: bool,

    /// Alias for --summary-only
    #[arg(short = 'q', long, action = ArgAction::SetTrue)]
    pub quiet: bool,

    /// Show all member responses, even if they fail
    #[arg(short = 'd', long, action = ArgAction::SetTrue)]
    pub verbose: bool,

    /// Re-enable all members
    #[arg(long, action = ArgAction::SetTrue)]
    pub all: bool,

    /// Enable codex (default: true)
    #[arg(long, action = ArgAction::SetTrue, default_value_t = true)]
    pub codex: bool,

    /// Disable codex
    #[arg(long = "no-codex", action = ArgAction::SetTrue)]
    pub no_codex: bool,

    /// Enable claude (default: true)
    #[arg(long, action = ArgAction::SetTrue, default_value_t = true)]
    pub claude: bool,

    /// Disable claude
    #[arg(long = "no-claude", action = ArgAction::SetTrue)]
    pub no_claude: bool,

    /// Enable gemini (default: true)
    #[arg(long, action = ArgAction::SetTrue, default_value_t = true)]
    pub gemini: bool,

    /// Disable gemini
    #[arg(long = "no-gemini", action = ArgAction::SetTrue)]
    pub no_gemini: bool,

    /// Ordered subset of codex,claude,gemini
    #[arg(long)]
    pub members: Option<String>,

    /// auto, codex, claude, or gemini
    #[arg(long)]
    pub summarizer: Option<String>,

    /// Reasoning effort: low, medium, or high
    #[arg(long)]
    pub effort: Option<String>,

    /// Color mode: auto, always, or never
    #[arg(long)]
    pub color: Option<String>,

    /// Disable color output
    #[arg(long = "no-color", action = ArgAction::SetTrue)]
    pub no_color: bool,

    /// Show version
    #[arg(short = 'v', long, action = ArgAction::SetTrue)]
    pub version: bool,

    /// Show help
    #[arg(short = 'h', long, action = ArgAction::SetTrue)]
    pub help_flag: bool,

    /// Query positional arguments
    #[arg(trailing_var_arg = true)]
    pub query_parts: Vec<String>,
}

#[derive(Debug, Clone)]
pub enum OutputMode {
    Text,
    Json,
    JsonStream,
}

#[derive(Debug, Clone)]
pub enum ColorMode {
    Auto,
    Always,
    Never,
}

#[derive(Debug, Clone)]
pub struct ParsedArgs {
    pub help: bool,
    pub version: bool,
    pub json: bool,
    pub json_stream: bool,
    pub headless: bool,
    pub plain: bool,
    pub no_banner: bool,
    pub summary_only: bool,
    pub quiet: bool,
    pub verbose: bool,
    pub color: ColorMode,
    pub summarizer: String,
    pub effort: Option<String>,
    pub timeout_ms: u64,
    pub max_member_chars: usize,
    pub cwd: String,
    pub members: Vec<String>,
    pub prompt_parts: Vec<String>,
}

pub fn parse_args(raw: RawArgs) -> Result<ParsedArgs, String> {
    if raw.help_flag {
        return Ok(ParsedArgs {
            help: true,
            version: false,
            members: ALL_ENGINES.iter().map(|s| s.to_string()).collect(),
            ..default_parsed()
        });
    }

    if raw.version {
        return Ok(ParsedArgs {
            help: false,
            version: true,
            members: ALL_ENGINES.iter().map(|s| s.to_string()).collect(),
            ..default_parsed()
        });
    }

    let color = if raw.no_color {
        ColorMode::Never
    } else {
        parse_color(raw.color.as_deref().unwrap_or("auto"))?
    };

    let summarizer = parse_summarizer(raw.summarizer.as_deref().unwrap_or(AUTO_SUMMARIZER))?;
    let effort = parse_effort(raw.effort.as_deref())?;

    let timeout_ms = match &raw.timeout {
        Some(v) => parse_timeout_ms(v)?,
        None => DEFAULT_TIMEOUT_MS,
    };

    let max_member_chars = match &raw.max_member_chars {
        Some(v) => parse_positive_integer(v, "--max-member-chars")?,
        None => DEFAULT_MAX_MEMBER_CHARS,
    };

    let cwd = raw.cwd.unwrap_or_else(|| {
        std::env::current_dir()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string()
    });

    let members = resolve_members(&raw)?;

    Ok(ParsedArgs {
        help: false,
        version: false,
        json: raw.json,
        json_stream: raw.json_stream || raw.ndjson,
        headless: raw.headless,
        plain: raw.plain,
        no_banner: raw.no_banner,
        summary_only: raw.summary_only,
        quiet: raw.quiet,
        verbose: raw.verbose,
        color,
        summarizer,
        effort,
        timeout_ms,
        max_member_chars,
        cwd,
        members,
        prompt_parts: raw.query_parts,
    })
}

fn resolve_members(raw: &RawArgs) -> Result<Vec<String>, String> {
    // --members wins over individual flags
    if let Some(members_str) = &raw.members {
        let members = parse_engine_list(members_str, "--members")?;
        if members.is_empty() {
            return Err("--members requires at least one engine.".to_string());
        }
        return Ok(members);
    }

    let mut enabled = std::collections::HashMap::new();
    for e in ALL_ENGINES {
        enabled.insert(*e, true);
    }

    if raw.all {
        for e in ALL_ENGINES {
            enabled.insert(*e, true);
        }
    }

    if raw.no_codex {
        enabled.insert("codex", false);
    }
    if raw.no_claude {
        enabled.insert("claude", false);
    }
    if raw.no_gemini {
        enabled.insert("gemini", false);
    }

    let members: Vec<String> = ALL_ENGINES
        .iter()
        .filter(|e| *enabled.get(*e).unwrap_or(&true))
        .map(|e| e.to_string())
        .collect();

    if members.is_empty() {
        return Err("At least one engine must be enabled.".to_string());
    }

    Ok(members)
}

fn parse_engine_list(value: &str, flag_name: &str) -> Result<Vec<String>, String> {
    let members: Vec<String> = value
        .split(',')
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
        .collect();

    if members.is_empty() {
        return Err(format!("{flag_name} requires at least one engine."));
    }

    let invalid: Vec<&str> = members
        .iter()
        .filter(|m| !ALL_ENGINES.contains(&m.as_str()))
        .map(|m| m.as_str())
        .collect();

    if !invalid.is_empty() {
        return Err(format!(
            "Unsupported engine in {flag_name}: {}",
            invalid.join(", ")
        ));
    }

    // Deduplicate while preserving order
    let mut seen = std::collections::HashSet::new();
    Ok(members.into_iter().filter(|m| seen.insert(m.clone())).collect())
}

fn parse_summarizer(value: &str) -> Result<String, String> {
    if value == AUTO_SUMMARIZER || ALL_ENGINES.contains(&value) {
        return Ok(value.to_string());
    }
    Err(format!("Unsupported summarizer: {value}"))
}

fn parse_effort(value: Option<&str>) -> Result<Option<String>, String> {
    match value {
        None | Some("") => Ok(None),
        Some(v) => {
            if EFFORT_LEVELS.contains(&v) {
                Ok(Some(v.to_string()))
            } else {
                Err(format!(
                    "Unsupported --effort value: {v} (expected {})",
                    EFFORT_LEVELS.join(", ")
                ))
            }
        }
    }
}

fn parse_timeout_ms(value: &str) -> Result<u64, String> {
    let seconds: f64 = value
        .parse()
        .map_err(|_| format!("Invalid timeout value: {value}"))?;

    if !seconds.is_finite() || seconds <= 0.0 {
        return Err(format!("Invalid timeout value: {value}"));
    }

    Ok((seconds * 1000.0).round() as u64)
}

fn parse_positive_integer(value: &str, flag_name: &str) -> Result<usize, String> {
    let parsed: usize = value
        .parse()
        .map_err(|_| format!("{flag_name} requires a positive integer."))?;

    if parsed == 0 {
        return Err(format!("{flag_name} requires a positive integer."));
    }

    Ok(parsed)
}

fn parse_color(value: &str) -> Result<ColorMode, String> {
    match value {
        "auto" => Ok(ColorMode::Auto),
        "always" => Ok(ColorMode::Always),
        "never" => Ok(ColorMode::Never),
        _ => Err(format!("Unsupported color mode: {value}")),
    }
}

fn default_parsed() -> ParsedArgs {
    ParsedArgs {
        help: false,
        version: false,
        json: false,
        json_stream: false,
        headless: false,
        plain: false,
        no_banner: false,
        summary_only: false,
        quiet: false,
        verbose: false,
        color: ColorMode::Auto,
        summarizer: AUTO_SUMMARIZER.to_string(),
        effort: None,
        timeout_ms: DEFAULT_TIMEOUT_MS,
        max_member_chars: DEFAULT_MAX_MEMBER_CHARS,
        cwd: std::env::current_dir()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string(),
        members: ALL_ENGINES.iter().map(|s| s.to_string()).collect(),
        prompt_parts: vec![],
    }
}

pub fn usage_text(version: &str) -> String {
    let default_timeout_seconds = DEFAULT_TIMEOUT_MS / 1_000;
    format!(
        r#"council v{version}

Ask multiple AI CLIs the same question, then synthesize their answers.

Usage:
  council [options] <query>
  echo "query" | council [options]
  council help

Examples:
  council "How should I structure this migration?"
  council --no-gemini --summarizer claude "Review this plan"
  council --headless --json "Summarize the implementation options"
  council --json-stream --codex --claude "Compare these designs"
  council --effort high "Analyze the tradeoffs for this architecture"

Selection:
  --members <list>          Ordered subset of codex,claude,gemini
  --codex / --no-codex      Enable or disable Codex
  --claude / --no-claude    Enable or disable Claude
  --gemini / --no-gemini    Enable or disable Gemini
  --all                     Re-enable all members
  --summarizer <name>       auto, codex, claude, or gemini

Output:
  --summary-only            Print only the final synthesis
  -q, --quiet               Alias for --summary-only
  -d, --verbose             Show all member responses, even if they fail
  --json                    Print structured JSON
  --json-stream             Stream JSONL lifecycle events
  --headless                Automation mode: no banner, no progress, summary-only text by default
  --plain                   Disable decorative UI and color
  --no-banner               Suppress the startup banner
  --color <auto|always|never>

Execution:
  --timeout <seconds>       Per-CLI timeout in seconds (default: {default_timeout_seconds} / 10 minutes)
  --max-member-chars <n>    Cap each member response before summarization (default: {DEFAULT_MAX_MEMBER_CHARS})
  --cwd <path>              Working directory for all upstream CLIs
  --effort <level>          Reasoning effort applied to every member: {effort_levels}

Other:
  -h, --help                Show help
  -v, --version             Show version

Environment:
  COUNCIL_CODEX_BIN         Override the codex executable path
  COUNCIL_CLAUDE_BIN        Override the claude executable path
  COUNCIL_GEMINI_BIN        Override the gemini executable path"#,
        version = version,
        default_timeout_seconds = default_timeout_seconds,
        effort_levels = EFFORT_LEVELS.join(", "),
    )
}
